core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -2690,7 +2691,14 @@ int complexity() {
     }
 
     @Override public RelDataType getType() {
-      return rexBuilder.typeFactory.leastRestrictive(Util.distinctList(types));
+      if (this.types.isEmpty()) {
+        // Expression is "x IS NULL"
+        return ref.getType();
+      }
+      final List<RelDataType> distinctTypes = Util.distinctList(this.types);
```

Comment:
I wonder if this can be integrated into `addRange`, so it skips adding duplicate items in the first place rather than add duplicates and de-duplicate it later.